### DISCUSSION TITLE
Audit logging of mutating changes to OpenFGA

### DISCUSSION THREAD

**Discussion by danielloader** [openfga/community#318] [Ideas]
Figured I'd kick this discussion off, I'm currently working on how to save an append only log of all the tuples added/removed to the openFGA system.

I've currently only got a fragile solution around wrapping the http calls in an audit log event which itself submits a record to a database but if either of those transactions fails I'm left with an incomplete or inaccurate audit log of what has happened.

Any ideas, solutions, or designs would be appreciated.

**Comment by ewanharris** [comment]
👋🏻 OpenFGA does keep a changelog of all tuples written that can be accessed via the [`/changes` API](https://openfga.dev/api/service#/Relationship%20Tuples/ReadChanges) with the relevant [`readChanges`](https://openfga.dev/docs/interacting/read-tuple-changes) function in an SDK.

If you were looking to maintain this outside of OpenFGA and improve reliability you could potentially look towards a pattern such as an [outbox pattern](https://microservices.io/patterns/data/transactional-outbox.html) or a [change data capture solution](https://www.confluent.io/learn/change-data-capture/) that would allow you to write to FGA and the append only log separately.

**Reply by salvadoriume** [reply]
Is it some feature that can be disabled?

**Reply by dyeam0** [reply]
Hello @salvadoriume, not sure what you mean by your question. The Read Changes API cannot be disabled.

**Reply by salvadoriume** [reply]
We do pretty lot of changes but we are not interested in changelog table, we don't use it and we wanted to disable funcionality that something will write to it. Possible?

**Reply by dyeam0** [reply]
It's not a "changelog" in the same sense of where a developer manually writes changes they want to log. It is a system-controlled log of tuple changes that are automatically written by OpenFGA every time you write or delete a tuple (ie. "tuple change log"). If you have no need for it, then you can safely ignore and not call it, but it cannot be disabled.

**Reply by salvadoriume** [reply]
That was my concern that if I cannot disable such feauture of writing to changelog table and I am not using it, now my table has 3mln rows and growing. So I need another, external process of cleaning this table. I very like openfga but there are cases when I need to do something additionaly (but it could be native solution). Another was when 3 pods are starting, checking for store_name, if not creating one. But finally there was 3 stores with different IDs as there is no lock mechanism for store_name so I needed to create index "unique_name_active_only" UNIQUE, btree (name) WHERE deleted_at IS NULL. Ok @dyeam0 thanks for reply and I will keep going and maintenance my external solutions :-)
